ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client/State.hs

```diff
@@ -0,0 +1,221 @@
+    -- see if they agree with the chain that the dynamo is serving; otherwise,
+    -- they become candidates to be the objector. See
+    -- 'ChainSyncJumpingJumperState' for more details.
+    Jumper
+      -- | A TVar containing the next jump to be executed.
+      !(StrictTVar m (Maybe (JumpInfo blk)))
+      -- | More precisely, the state of the jumper.
+      !(ChainSyncJumpingJumperState blk)
+  deriving (Generic)
+
+deriving anyclass instance
+  ( IOLike m,
+    HasHeader blk,
+    LedgerSupportsProtocol blk,
+    NoThunks (Header blk)
+  ) => NoThunks (ChainSyncJumpingState m blk)
+
+-- | The ChainSync state required for jumps
+--
+-- The jump info is mostly a snapshot of the @KnownIntersectionState@ of the
+-- dynamo, with the difference that 'jTheirFragment' might be a proper prefix of
+-- the original candidate fragment.
+--
+-- This can happen if we need to look for an intersection when the jumper
+-- rejects a jump.
+data JumpInfo blk = JumpInfo
+  { jMostRecentIntersection  :: !(Point blk)
+  , jOurFragment             :: !(AnchoredFragment (Header blk))
+  , jTheirFragment           :: !(AnchoredFragment (Header blk))
+  , jTheirHeaderStateHistory :: !(HeaderStateHistory blk)
+  }
+  deriving (Generic)
+
+instance (HasHeader (Header blk)) => Eq (JumpInfo blk) where
+  (==) = (==) `on` headPoint . jTheirFragment
+
+instance LedgerSupportsProtocol blk => NoThunks (JumpInfo blk) where
+  showTypeOf _ = show $ typeRep (Proxy @(JumpInfo blk))
+
+-- | The specific state of a jumper peer. This state is to be understood as “to
+-- the best of our knowledge”, that is “last time we asked them”. For instance,
+-- a jumper might be marked as 'Happy' even though its chain has been differing
+-- from the dynamo's for hundreds of blocks, if we haven't asked them to jump
+-- since then.
+data ChainSyncJumpingJumperState blk
+  = -- | The jumper is happy with the dynamo, and we hold the jump info of the
+    -- last accepted jump.
+    Happy JumperInitState !(Maybe (JumpInfo blk))
+  | -- | The jumper disagrees with the dynamo and we are searching where exactly
+    -- that happens. All we know is a point where the jumper agrees with the
+    -- dynamo and a point where the jumper disagrees with the dynamo, carried by
+    -- this constructor.
+    --
+    -- INVARIANT: The tip of the fragment in the good jump info (first argument)
+    -- is in the fragment of the bad jump info or is an ancestor of it.
+    LookingForIntersection !(JumpInfo blk) !(JumpInfo blk)
+  | -- | The jumper disagrees with the dynamo and we have determined the latest
+    -- point where dynamo and jumper agree. We store here the jump info of the
+    -- latest accepted jump and the point of the earliest rejected jump.
+    --
+    -- The init state indicates the initialization to use for the objector in
+    -- case this jumper is promoted.
+    FoundIntersection ObjectorInitState !(JumpInfo blk) !(Point (Header blk))
```

Comment:
I wonder whether these should still be jumpers or there should be in some other state. Such as, once a peer becomes `LookingForIntersection` how can it recover back to `Happy`?

I wonder if these should be some other `Pre-Objector` state.

Reply:
A peer LookingForIntersection becomes Happy when a new dynamo is elected. This is transition `h` in the state diagram.